pkg/commands/copy_test.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2018 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package commands
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/GoogleContainerTools/kaniko/pkg/dockerfile"
+	"github.com/GoogleContainerTools/kaniko/testutil"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/moby/buildkit/frontend/dockerfile/instructions"
+	"github.com/sirupsen/logrus"
+)
+
+var copyTests = []struct {
+	name           string
+	SourcesAndDest []string
+	expectedDest   []string
+}{
+	{
+		name:           "copy foo into tempCopyExecuteTest/",
+		SourcesAndDest: []string{"foo", "tempCopyExecuteTest/"},
+		expectedDest:   []string{"foo"},
+	},
+	{
+		name:           "copy foo into tempCopyExecuteTest",
+		SourcesAndDest: []string{"foo", "tempCopyExecuteTest"},
+		expectedDest:   []string{"tempCopyExecuteTest"},
+	},
+}
+
+func TestCopyExecuteCmd(t *testing.T) {
+	cfg := &v1.Config{
+		Cmd:        nil,
+		Env:        []string{},
+		WorkingDir: "../../integration/context/",
+	}
+
+	for _, test := range copyTests {
+		t.Run(test.name, func(t *testing.T) {
+			dirList := []string{}
+
+			logrus.Infof("Running test: %v", test.name)
+
+			cmd := CopyCommand{
+				cmd: &instructions.CopyCommand{
+					SourcesAndDest: test.SourcesAndDest,
+				},
+				buildcontext: "../../integration/context/",
+			}
+
+			buildArgs := copySetUpBuildArgs()
+			dest := cfg.WorkingDir + test.SourcesAndDest[len(test.SourcesAndDest)-1]
```

Comment:
Is there a reason we can't use `tempDir` ? The advantage of creating a temp and writing all the files in the temp are
1. We can delete the who dir in defer. In case the test got cancelled or killed, `defer` block is always executed and dirs are cleaned up.
```
dir, err := ioutil.TempDir("dir", "prefix")
if err != nil {
       t.Errorf("error creating temp dir %s", err)
	t.Fail()
}
defer os.RemoveAll(dir)
```

Reply:
Okay, reworked the whole thing.